### PR TITLE
Add TCP support for GELF log driver

### LINKED ISSUE #29871: Unable to use tcp in --gelf-address
Hello,

We are running our application in AWS ECS and we are trying to collect our container logs via TCP port using gelf log driver. But if I used TCP in --log-driver-options --gelf-address tcp://ipaddress:port my container is not starting.

docker: Error response from daemon: gelf: endpoint needs to be UDP.

Kindly suggest how to use TCP in gelf-address option.

My docker version is 1.12

Regards,
Raja

### LINKED ISSUE #33495: GELF Driver does not support TCP
**Description**

The GELF Log driver does not currently support the TCP transport. In some cases, TCP would be more appropriate. 

It is currently not supported directly by go-gelf, but there is currently a pull request (https://github.com/Graylog2/go-gelf/pull/9) that adds TCP support. 

I am currently working on a pull request that will enable the TCP transport once the go-gelf PR is merged.

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
**- What I did**
I added TCP support to the GELF log driver. closes #33495 closes #29871

The documentation PR can be found [here](https://github.com/docker/docker.github.io/pull/4536)

**- How I did it**
The following changes were done: 

- Updated go-gelf to v2
- Added unit tests to the gelf log driver
- Refactored the gelf log driver to make room for the TCP support
- Added TCP support

**- How to verify it**
1. Run a Graylog server and configure a TCP GELF Input and a UDP GELF Input. 

2. Run a container with: 
  - "--log-driver gelf --log-opt gelf-address=udp://<graylog_host>:<udp_port>"
  - "--log-driver gelf --log-opt gelf-address=tcp://<graylog_host>:<tcp_port>"

3. Validate that the container logs are received in the Graylog server.

**- Description for the changelog**
Added TCP support to the GELF log driver.

**- A picture of a cute animal (not mandatory but encouraged)**
animal

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @vieux @cpuguy83 @mariussturm PTAL

**Comment by vieux** [comment]
LGTM

**Review by boaz0 — COMMENTED** [review]
LGTM

**Comment by thaJeztah** [comment]
ping @mariussturm could you have a look at this one?

**Comment by mariussturm** [comment]
LGTM

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Thanks for reviewing @mariussturm

**Comment by pdepaepe** [comment]
Thanks a lot for this PR. Would you please also consider tcp+tls:// ?

**Comment by ghislainbourgeois** [comment]
@pdepaepe I will look into what is involved, but do not expect it too quickly. This will need a PR in the go-gelf library before making the PR in moby. From my experience with this one, this could take a couple of months.

File: vendor.conf [hunk 1]
```diff
@@ -76,7 +76,7 @@ github.com/syndtr/gocapability 2c00daeb6c3b45114c80ac44119e7b8801fdd852
 github.com/golang/protobuf 7a211bcf3bce0e3f1d74f9894916e6f116ae83b4
 
 # gelf logging driver deps
-github.com/Graylog2/go-gelf 7029da823dad4ef3a876df61065156acb703b2ea
+github.com/Graylog2/go-gelf v2
```
Comment (thaJeztah) [inline vendor.conf]: Whoops, turned out `v2` here is a branch, not a tag

**Comment by Hermain** [comment]
@ghislainbourgeois I'm trying to send logs to logstash. My question: Which delimiter is used to separate entries? Is it \0? Should this maybe be configurable so that \n could be used (as this is the default for most log parsers)

**Pull request merged**